File: src/main/java/org/featx/jvmon/model/PageResponse.java
package org.featx.jvmon.model;

import lombok.Data;
import lombok.EqualsAndHashCode;
import org.featx.jvmon.exception.ErrorCode;

import java.util.List;

/**
 * @author Excepts
 * @since 2019/10/12 11:12
 */
@Data
@EqualsAndHashCode(callSuper = true)
public class PageResponse<R> extends ListResponse<R> {

    private Integer total;

    private Integer page;

    public PageResponse<R> total(Integer total) {
        this.total = total;
        return this;
    }

    public PageResponse<R> page(Integer page) {
        this.page = page;
        return this;
    }

    public PageResponse<R> with(int page, int total) {
        this.total = total;
        this.page = page;
        return this;
    }

    private PageResponse(int code, String message, List<R> result) {
        this.setCode(code);
        this.setMessage(message);
        this.setResult(result);
    }

    public static <R> PageResponse<R> succeeded(List<R> result)  {
        return new PageResponse<>(ErrorCode.NO_ERROR, null, result);
    }
}

File: src/main/kotlin/org/featx/jvmon/model/PageResponse.kt
package org.featx.jvmon.model

import lombok.Data
import lombok.EqualsAndHashCode
import org.featx.jvmon.exception.BusinessError
import org.featx.jvmon.exception.ErrorCode

/**
 * @author Excepts
 * @since 2019/10/12 11:12
 */
@Data
@EqualsAndHashCode(callSuper = true)
class PageResponse<T> private constructor(code: Int, msg: String?, result: List<T>?) :
        ListResponse<T>(code, msg, result) {

    var page: Int? = null
        set(page) {
            field = this.page
        }

    var total: Int? = null
        set(total) {
            field = this.total
        }

    companion object {
        fun <R> succeeded(result: List<R>): PageResponse<R> {
            return PageResponse(ErrorCode.NO_ERROR, null, result)
        }

        fun <R> occur(businessError: BusinessError): PageResponse<R> {
            return PageResponse(businessError.code, businessError.message, null)
        }

        fun <R> occur(businessError: BusinessError, extra: List<R>): PageResponse<R> {
            return PageResponse(businessError.code, businessError.message, extra)
        }
    }
}
